File: Ex15.java
package Ex15;

class SaldoInsuficienteException extends Exception {
    public SaldoInsuficienteException(String mensagem) {
        super(mensagem);
    }
}

class ContaBancaria {
    private String titular;
    private double saldo;

    public ContaBancaria(String titular, double saldoInicial) {
        this.titular = titular;
        this.saldo = saldoInicial;
    }

    public void depositar(double valor) {
        saldo += valor;
        System.out.println("Deposito de R$ " + valor + " realizado com sucesso.");
    }

    public void sacar(double valor) throws SaldoInsuficienteException {
        if (valor > saldo) {
            throw new SaldoInsuficienteException("Saldo insuficiente para realizar o saque de R$ " + valor);
        }
        saldo -= valor;
        System.out.println("Saque de R$ " + valor + " realizado com sucesso.");
    }

    public double getSaldo() {
        return saldo;
    }

    public String getTitular() {
        return titular;
    }
}

public class Ex15 {
    public static void main(String[] args) {
        ContaBancaria conta = new ContaBancaria("Alice", 1000.0);
        
        try {
            conta.sacar(200.0);
            System.out.println("Saldo atual: R$ " + conta.getSaldo());

            conta.sacar(900.0); 
        } catch (SaldoInsuficienteException e) {
            System.out.println("Erro: " + e.getMessage());
        }
    }
}

File: ex15.go
package main

import (
	"fmt"
)

type SaldoInsuficienteException struct {
	mensagem string
}

func (e *SaldoInsuficienteException) Error() string {
	return e.mensagem
}

type ContaBancaria struct {
	saldo float64
}

func (c *ContaBancaria) Sacar(valor float64) error {
	if valor > c.saldo {
		return &SaldoInsuficienteException{mensagem: "Saldo insuficiente para o saque"}
	}
	c.saldo -= valor
	return nil
}

func (c *ContaBancaria) Depositar(valor float64) {
	c.saldo += valor
}

func main() {
	conta := ContaBancaria{saldo: 100.0}

	err := conta.Sacar(150.0)
	if err != nil {
		fmt.Println("Erro:", err)
	} else {
		fmt.Println("Saque realizado com sucesso!")
	}

	conta.Depositar(50.0)
	fmt.Println("Saldo atual:", conta.saldo)

	err = conta.Sacar(100.0)
	if err != nil {
		fmt.Println("Erro:", err)
	} else {
		fmt.Println("Saque realizado com sucesso!")
	}
}
